lib/email-addresses.js

```diff
@@ -782,19 +796,35 @@ function parse5322(opts) {
         return result;
     }
 
-    function giveResult(ast) {
+    function _giveResult (ast, types) {
         function grabSemantic(n) {
             return n !== null ? n.semantic : null;
-        }
-        var i, ret, addresses, addr, name, aspec, local, domain, comments;
+        }        
+        var i, ret, addresses, addr, name, aspec, local, domain, comments, group_addresses;
         if (ast === null) {
             return null;
         }
         ret = { ast: ast };
-        addresses = findAllNodes('address', ast);
+        addresses = [].concat.apply([], types.map(function (type) { return findAllNodes(type, ast) }));
         ret.addresses = [];
         for (i = 0; i < addresses.length; i += 1) {
             addr = addresses[i];
+            if (addr.seen) continue;
+            addr.seen = true;
+            if (addr.name === 'group') {
```

Comment:
It happens to work, because `findAllNodes` is first called on `group` and then on `mailbox`, but considering this loop by itself one might wonder if a node in a `group` could be processed before the `group` it is part of. If you called `_giveResult(ast, ['mailbox', 'group'])` somewhere else, you would get the wrong result, which is confusing.

I think it would be simpler to:
1) make this two functions (no recursion)
2) process all the 'group' nodes first (loop over `findAllNodes('group', ast)`, then process non group nodes underneath that group, marking seen)
3) process all the non-group nodes second (loop over `findAllNodes('mailbox', ast)`, skipping if "seen")